### ISSUE TITLE
Replace lists with `NonEmpty` - or, do we allow no-field records?

### ISSUE THREAD

**Issue by parsonsmatt** [parsonsmatt/prairie#7]
There are a few types that use `[]` instead of `NonEmpty`, which (while convenient) restrict the utility.

For example, `allFields` is `allFields :: (Record rec) => [SomeField rec]`. However, if we don't allow empty `Record`s, then we can write that as `NonEmpty (SomeField rec)`, which is much nicer - we can use `sconcat` and `Semigroup` classes generally.

**Comment by prikhi** [comment]
This would be sick, my usecase is wanting to safely `foldMapRecord` into a NonEmpty list. Even if we just had the TH generate some `PraireNonEmptyFieldList` method-less instance that I could use to gate a `fromJust . nonEmpty` call.

**Comment by parsonsmatt** [comment]
Yeah. I think the main question here is: do we want to allow something like:

```haskell
data Dumb = Dumb

mkRecord ''Dumb
```

I think no - that's silly - and so we should use something like `Foldable1` for these. But that may make interfaces more irritating.
